### PR TITLE
Rewrote alter_index

### PR DESCRIPTION
**Pull request by madams51703**

### CODE REVIEW

File: plsql/PlSqlParser.g4 [hunk 1]
```diff
@@ -1058,6 +1184,33 @@ create_mv_refresh
...
+//https://docs.oracle.com/cd/E11882_01/server.112/e41084/statements_5001.htm#SQLRF01201
+create_cluster
+    : CREATE CLUSTER  cluster_name '(' (','? column_name datatype SORT?)+ ')'
+          ((physical_attributes_clause
+           | SIZE size_clause
+           | TABLESPACE tablespace
+           | INDEX
+           | (SINGLE TABLE)? HASHKEYS UNSIGNED_INTEGER (HASH IS expression)?
+           )+
+          )?
```
Comment (KvanTTT) [inline plsql/PlSqlParser.g4]: I think it can be rewritten with the star syntax: `()*` instead of `(()+)?`.

File: plsql/PlSqlParser.g4 [hunk 2]
```diff
@@ -659,9 +654,140 @@ odci_parameters
 indextype
     : (id_expression '.')? id_expression
     ;
-  
+
+//https://docs.oracle.com/cd/E11882_01/server.112/e41084/statements_1010.htm#SQLRF00805  
 alter_index
-    : ALTER INDEX old_index_name=index_name RENAME TO new_index_name=index_name ';'
+    : ALTER INDEX index_name (alter_index_ops_set1 | alter_index_ops_set2)
+      ';'
```
Comment (KvanTTT) [inline plsql/PlSqlParser.g4]: A semicolon can be moved to the end of the previous line.

File: plsql/PlSqlParser.g4 [hunk 3]
```diff
@@ -659,9 +654,140 @@ odci_parameters
...
+    ;
+
+alter_index_ops_set1
+    : (deallocate_unused_clause
```
Comment (KvanTTT) [inline plsql/PlSqlParser.g4]: Please add a space after left parenthesis.

File: plsql/PlSqlParser.g4 [hunk 4]
```diff
@@ -659,9 +654,140 @@ odci_parameters
...
+      )+
+    ;
+
+alter_index_ops_set2
+    : rebuild_clause
+    | PARAMETERS '(' odci_parameters ')'
+    | COMPILE
+    | enable_or_disable
+    | UNUSABLE
+    | (VISIBLE | INVISIBLE)
```
Comment (KvanTTT) [inline plsql/PlSqlParser.g4]: Maybe use separated parse rule similar to other `ENABLE | DISABLE` and other?

File: plsql/PlSqlParser.g4 [hunk 5]
```diff
@@ -659,9 +654,140 @@ odci_parameters
...
+              (( parallel_clause
+               | TABLESPACE tablespace
+               | PARAMETERS '(' odci_parameters ')'
+//TODO         | xmlindex_parameters_clause
+               | ONLINE
+               | physical_attributes_clause
+               | key_compression
+               | logging_clause
+               )+
+              )?
```
Comment (KvanTTT) [inline plsql/PlSqlParser.g4]: `(()+)?` can be replaced with the star syntax: `()*`.

File: plsql/PlSqlParser.g4 [hunk 6]
```diff
@@ -659,9 +654,140 @@ odci_parameters
...
+        key_compression? parallel_clause?
+    ;
+
+coalesce_index_partition
+    : COALESCE PARTITION parallel_clause?
+    ;
+
+modify_index_partition
+    : MODIFY PARTITION partition_name
+        (modify_index_partitions_ops+
```
Comment (KvanTTT) [inline plsql/PlSqlParser.g4]: Space after the left parenthesis.

**Comment by teverett** [comment]
Try this, it's far easier.

https://github.com/antlr/Antlr4Formatter

**Comment by madams51703** [comment]
Thanks for sharing Tom!!!

On Dec 15, 2017 10:39 AM, "Tom Everett" <notifications@github.com> wrote:

Try this, it's far easier.

https://github.com/antlr/Antlr4Formatter

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/antlr/grammars-v4/pull/1015#issuecomment-352036347>,
or mute
the thread

**Comment by madams51703** [comment]
Changes made.  Thanks for clarifying the space/no space after (

**Pull request merged**